nebula_decoders/include/nebula_decoders/nebula_decoders_velodyne/decoders/velodyne_generic_decoder.hpp

```diff
@@ -0,0 +1,445 @@
+#pragma once
+
+#include "nebula_decoders/nebula_decoders_velodyne/decoders/velodyne_scan_decoder.hpp"
+#include "nebula_decoders/nebula_decoders_velodyne/decoders/vlp_16.hpp"
+#include "nebula_decoders/nebula_decoders_velodyne/decoders/vls_128.hpp"
+
+#include <velodyne_msgs/msg/velodyne_packet.hpp>
+#include <velodyne_msgs/msg/velodyne_scan.hpp>
+
+#include <angles/angles.h>
+
+#include <array>
+#include <cmath>
+#include <utility>
+
+namespace nebula
+{
+namespace drivers
+{
+
+// vlp16 specific constant
+constexpr uint32_t MAX_POINTS = 300000;
+
+/// @brief Velodyne LiDAR decoder
+template <typename SensorT>
+class VelodyneDecoder : public VelodyneScanDecoder
+{
+public:
+  // /// @brief Constructor
+  // /// @param sensor_configuration SensorConfiguration for this decoder
+  // /// @param calibration_configuration Calibration for this decoder
+  // explicit VelodyneDecoder(
+  //   const std::shared_ptr<drivers::VelodyneSensorConfiguration> & sensor_configuration,
+  //   const std::shared_ptr<drivers::VelodyneCalibrationConfiguration> &
+  //   calibration_configuration);
+  // /// @brief Parsing and shaping VelodynePacket
+  // /// @param velodyne_packet
+  // void unpack(const velodyne_msgs::msg::VelodynePacket & velodyne_packet) override;
+  // /// @brief Get the flag indicating whether one cycle is ready
+  // /// @return Readied
+  // bool hasScanned() override;
+  // /// @brief Calculation of points in each packet
+  // /// @return # of points
+  // int pointsPerPacket() override;
+  // /// @brief Get the constructed point cloud
+  // /// @return tuple of Point cloud and timestamp
+  // std::tuple<drivers::NebulaPointCloudPtr, double> get_pointcloud() override;
+  // /// @brief Resetting point cloud buffer
+  // /// @param n_pts # of points
+  // void reset_pointcloud(size_t n_pts, double time_stamp) override;
+  // /// @brief Resetting overflowed point cloud buffer
+  // void reset_overflow(double time_stamp) override;
+
+  explicit VelodyneDecoder(
+    const std::shared_ptr<drivers::VelodyneSensorConfiguration> & sensor_configuration,
+    const std::shared_ptr<drivers::VelodyneCalibrationConfiguration> & calibration_configuration)
+  {
+    sensor_configuration_ = sensor_configuration;
+    calibration_configuration_ = calibration_configuration;
+
+    scan_timestamp_ = -1;
+
+    scan_pc_.reset(new NebulaPointCloud);
+    overflow_pc_.reset(new NebulaPointCloud);
+
+    // Set up cached values for sin and cos of all the possible headings
+    for (uint16_t rot_index = 0; rot_index < ROTATION_MAX_UNITS; ++rot_index) {
+      double rotation = angles::from_degrees(ROTATION_RESOLUTION * rot_index);
+      rotation_radians_[rot_index] = rotation;
+      cos_rot_table_[rot_index] = cosf(rotation);
+      sin_rot_table_[rot_index] = sinf(rotation);
+    }
+
+    phase_ = (uint16_t)round(sensor_configuration_->scan_phase * 100);
```

Comment:
Please use `static_cast<uint16_t>(...)` in favor of C-style `(uint16_t)(...)` casts. Static casts only compile if the cast is actually safe to do, while C-style casts can do a lot of weird stuff. In this case it should be fine, but please still change it.